script/heatmap.py

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python3
+#
+# Parse the profiling info that is produced with --smtprof and produce a heatmap.
+# If this script is too slow, we should use the hierarchical structure of the code
+# regions. For now, the code is using sorted lists instead.
+#
+# Igor Konnov, 2020
+
+import csv
+import html
+import itertools
+import re
+import sys
+
+class ProfileEntry:
+    """Profiling data"""
+    def __init__(self, weight, ncells, nconsts, nexprs):
+        self.weight = weight
+        self.ncells = ncells
+        self.nconsts = nconsts
+        self.nexprs = nexprs
+
+    def add(self, other):
+        self.weight += other.weight
+        self.ncells += other.ncells
+        self.nconsts += other.nconsts
+        self.nexprs += other.nexprs
+
+class CodeRegion:
+    """An hierarchical code region that carries profiling data"""
+
+    def __init__(self, filename, fromPos, toPos, profEntry):
+        self.filename = filename
+        self.fromPos = fromPos
+        self.toPos = toPos
+        self.profEntry = profEntry
+        # sort by (end_line, start_line)
+        self.sort_key = (toPos[0], fromPos[0])
+
+    def rangeByLine(self, line_no, width):
+        if self.fromPos[0] > line_no or self.toPos[0] < line_no:
+            return (-1, -1)
+
+        if self.fromPos[0] < line_no:
+            if self.toPos[0] > line_no:
+                # the region spans the whole line
+                return (1, width)
+            else:
+                # the region starts on a preceding line but end on this one
+                return (1, self.toPos[1])
+        else:
+            if self.toPos[0] > line_no:
+                # the region starts on this line but ends later
+                return (self.fromPos[1], width)
+            else:
+                # the region starts and ends on this line
+                return (self.fromPos[1], self.toPos[1])
+
+    def isInside(self, other):
+        self.filename = other.filename \
+                and self.fromPos >= outer.fromPos \
+                and self.toPos <= outer.toPos
+
+
+# the regular expression for the source location as produced by SANY
+REGION_RE = re.compile(r"(.*):(\d+):(\d+)-(\d+):(\d+)")
+
+def region_from_csv(row):
+    m = REGION_RE.match(row[4])
+    if m:
+        entry = ProfileEntry(int(row[0]), int(row[1]), int(row[2]), int(row[3]))
+        return CodeRegion(m.group(1),
+                      (int(m.group(2)), int(m.group(3))), 
+                      (int(m.group(4)), int(m.group(5))),
```

Comment:
No essential, but I think you could improve readability here by using named groups in the regex and a `gruopdict` here: https://docs.python.org/3/library/re.html#re.Match.groupdict

Reply:
Yeah. I think this script has to be reviewed at some point and simplified. I just wanted to try the idea. Dealing with code regions happened to be hard.